framework/include/executors/Executor.h

```diff
@@ -0,0 +1,166 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Executioner.h"
+#include "ExecutorInterface.h"
+
+#include <string>
+
+class Problem;
+class Executor;
+
+/// The Executor class directs the execution flow of simulations.  It manages
+/// outputting, time stepping, mesh adaptivity, solve sequencing, multiapp
+/// execution, etc.  Users can compose executor objects in an input file using
+/// the same pattern as with mesh generators.  Subclass here and implement your
+/// own run function if you need special or fancy functionality that
+/// isn't supported by the default executors available.
+class Executor : public Executioner, public ExecutorInterface
+{
+public:
+  /// This object tracks the success/failure state of the executor system as
+  /// execution proceeds in a simulation.  Because executors can be composed into
+  /// trees, result objects are correspondingly composed into trees to track
+  /// simulation progress.  Result objects should generally be created by
+  /// executors calling the Result::newResult() function rather than by using
+  /// the Result constructor directly.
+  struct Result
+  {
+    Result() : converged(true), _name("NO_NAME") {}
+    Result(const std::string & name) : converged(true), _name(name) {}
+    Result(const MooseObject * obj) : converged(true), _name(obj->name()) {}
+
+    /// whether or not a executor ran its code successfully - only reports
+    /// results from the executor itself.  If a sub/internal executor of a executor
+    /// fails, that sub-executor's result object will have converged=false, while
+    /// the parent may still have converged=truee. Users should use convergedAll
+    /// to recursively determine if there was any descendant executor failure.
+    bool converged = true;
+
+    /// Optional message detailing why a executor passed or failed (i.e. failed to converge).
+    std::string reason;
+
+    /// Maps a name/label of a executor's internal/sub executors to the result
+    /// object returned by running each of those internal/sub executors.  This
+    /// member should generally not be accessed directly.  It should generally
+    /// be populated through the record function.  Info contained in these
+    /// results will be included in printouts from the str function.
+    std::map<std::string, Result> subs;
+
+    /// Prints a full recursive output of this result object - including all
+    /// descendant's results.  If success_msg is true, then all result output
+    /// that contains a message will be printed even if it converged/passed.
```

Comment:
```suggestion
    /// descendant's results.  If success_msg is true, then all results' output
    /// that contain a message will be printed even if it converged/passed.
```

Reply:
I think the original text is more clear.